src/contrib/cluster/Akka.Cluster.Tools/Singleton/OldestChangedBuffer.cs

```diff
@@ -135,65 +126,64 @@ private void SetupCoordinatedShutdown()
-        }
-
-        private bool MatchingRole(Member member)
-        {
-            return string.IsNullOrEmpty(_role) || member.HasRole(_role);
-        }
-
-        private void HandleInitial(ClusterEvent.CurrentClusterState state)
+        
+        private void HandleInitial(ClusterEvent.CurrentClusterState clusterState)
         {
             // all members except Joining and WeaklyUp
-            _membersByAge = state.Members
-                .Where(m => m.UpNumber != int.MaxValue && MatchingRole(m))
+            var newMembersByAge = clusterState.Members
+                .Where(m => m.UpNumber != int.MaxValue && State.MatchingRole(m))
                 .ToImmutableSortedSet(_memberAgeComparer);
+            
+           State = State with { MembersByAge = newMembersByAge };
+           
+           // compute the initial oldest - doesn't matter if it's not "safe" or not. That's our parent's problem.
+           // The subsequent `MemberRemoved` emits that will be emitted if an older node is leaving or downed
+           // will clean that up inside our state.
+           var (newState, oldestChanged) = State.ComputeNextOldest();
+           State = newState;
 
             // If there is some removal in progress of an older node it's not safe to immediately become oldest,
             // removal of younger nodes doesn't matter. Note that it can also be started via restart after
             // ClusterSingletonManagerIsStuck.
-            var selfUpNumber = state.Members
+            var selfUpNumber = clusterState.Members
                 .Where(m => m.UniqueAddress == _cluster.SelfUniqueAddress)
                 .Select(m => (int?)m.UpNumber)
                 .FirstOrDefault() ?? int.MaxValue;
 
-            var oldest = _membersByAge.TakeWhile(m => m.UpNumber <= selfUpNumber).ToList();
+            var oldest = State.MembersByAge.TakeWhile(m => m.UpNumber <= selfUpNumber).ToList();
             var safeToBeOldest = !oldest.Any(m => m.Status is MemberStatus.Down or MemberStatus.Exiting or MemberStatus.Leaving);
             var initial = new InitialOldestState(oldest.Select(m => m.UniqueAddress).ToList(), safeToBeOldest);
             _changes = _changes.Enqueue(initial);
         }
 
         private void Add(Member member)
         {
-            if (MatchingRole(member))
-                TrackChanges(() =>
+            if (State.MatchingRole(member))
+            {
+                var (newState, oldestChanged) = State.AddMember(member);
+                State = newState;
+                if (oldestChanged)
                 {
-                    // replace, it's possible that the upNumber is changed
-                    _membersByAge = _membersByAge.Remove(member);
-                    _membersByAge = _membersByAge.Add(member);
-                });
+                    _changes = _changes.Enqueue(new OldestChanged(State.CurrentOldest?.UniqueAddress));
```

Comment:
Same behavior as before, but this is just now internally handled by the `State`.